### ISSUE TITLE
(Request) Advertise support for IntelliJ IDEA

### ISSUE THREAD

**Issue by brittyazel** [Ketho/vscode-wow-api#143]
I am a long-time user of Intellij IDEA with the EmmyLUA plugin, and after maintaining my own API docs for many years, [here](https://github.com/Ellypse/IntelliJ-IDEA-Lua-IDE-WoW-API/), I have now been using this project as it is far superior and far better maintained. Interestingly, this project, while advertised as being for VSCode specifically, works essentially perfect out-of-the-box for IntelliJ IDEA and EmmyLUA. Given that, perhaps you might consider adding IntelliJ IDEA to the advertised list of IDEs.

That being said, I know the name of the project itself has "vscode-" prefixing it, so I wouldn't request you alter that at all, but perhaps in the README it might be worth mentioning the working IntelliJ IDEA / EmmyLUA support for those like me that are developing using those alternative tools. Either way, accidental or not, your repo is hands-down the best in-IDE implementation of the WoW API Documentation for both VSCode _and_ IntelliJ IDEA, and I say that confidently as someone who, until recently, developed a far inferior competitor to this project lol.

**Comment by Ketho** [comment]
I actually have no idea how this project works for IntelliJ, is there something I should change to better target other IDEs like IntelliJ?

I do know some people somehow use at least my annotations for IntelliJ, but not their exact setup and if they use the same language server or a different one. I'm open to working together with you on this.

**Comment by brittyazel** [comment]
It seems to work perfectly fine in IntelliJ as far as I can tell, granted I haven't done an in-depth side-by-side between VSCode and IntelliJ. I think the big ticket item here is EmmyLUA, which supports both VSCode and IntelliJ as a native plugin. I think, for the most part, as long as this project supports EmmyLUA's format considerations it should just 'work' for both. I'm not really sure what else you could add to support IntelliJ 'more' tbh. There might be some minor differences in the mouseover tooltips and how they might handle rich-text features, but I'd just be guessing.

Just so you can see, this is how I import your docs into my project:
![ProjectStructure](https://github.com/Ketho/vscode-wow-api/assets/3747347/80bf6458-1aff-48ab-84ad-27695d58bd5d)

And it shows up in my sidebar as this:
![Sidebar](https://github.com/Ketho/vscode-wow-api/assets/3747347/af847f6e-dad2-42dd-943f-789a5fde85cf)

And tooltips end up looking like this:
![Tooltip](https://github.com/Ketho/vscode-wow-api/assets/3747347/4f12219a-7dd7-44f6-847a-678a72b01f13)

**Comment by Ketho** [comment]
So if I understand this correct, for IntelliJ documentation it should say that users should get [EmmyLua](https://plugins.jetbrains.com/plugin/9768-emmylua), clone/download this repo and add `../vscode-wow-api/EmmyLua` to the Lua path?

I went ahead added you as a collaborator if you want to add that to the README.md or somewhere else.

**Comment by brittyazel** [comment]
Oh, thank you for adding me! I'll make sure to not abuse the privilege lol.

I believe a use could add the root vscode-wow-api folder as a Classpath, and it would work just as well, but by focusing it do the EmmyLUA folder it avoids unnecessarily indexing some of the supporting utilities and focuses on just the docs themselves. But otherwise, yeah. Once you add the classpath it goes ahead and indexes the files starts working immediately.

**Comment by brittyazel** [comment]
Yeah for things like Enums and CVars I'm not really sure how best to handle that. I tried to come up with a solution [myself](https://github.com/Ellypse/IntelliJ-IDEA-Lua-IDE-WoW-API/blob/master/APIs/EnumValues.lua) using raw EmmyLUA features (at least from what I could figure out), but it was never totally satisfactory. It would sometimes provide suggestions and other times not.

There is likely some hidden EmmyLUA functionality for exactly this that I was ignorant to, but my main focus was just getting parsing from warcraft.wiki.gg and the api source code working broadly.

**Comment by BreakBB** [comment]
Maybe it is more IDE agnostic to bundle the type annotations as [LLS Addon](https://github.com/LuaLS/LLS-Addons) and distribute it as this? 🤔

I am using https://plugins.jetbrains.com/plugin/22315-sumnekolua for IntelliJ and added the [Annotations](https://github.com/Ketho/vscode-wow-api/tree/master/Annotations) to the `.luarc.json`:

```json
{
    ...
    "workspace.userThirdParty": [".types/LuaAddons"],
}
```

To mark it LLS Addon all it needs is the correct directory structure. As example:

```
└── wow-api
    ├── config.json
    └── library
        ├── Data
        │   ├── CVar.lua
        │   ├── ...
        ├── Interface
        │   ├── ...
        ├── Libraries
        │   ├── ...
        ├── Lua
        │   ├── ...
        ├── Type
        │   ├── ...
        ├── Widget
        │   ├── ...
```

**Comment by Ketho** [comment]
> Maybe it is more IDE agnostic to bundle the type annotations as [LLS Addon](https://github.com/LuaLS/LLS-Addons) and distribute it as this? 🤔

I'm certainly entertaining the idea of bundling it as a LLS Addon for IDE agnostic support. I do need to find out how the vscode 3rd party extension (which has some extra features) would not conflict with the LLS addon

**Comment by Thaodan** [comment]
I hope this fits in here but would it be possible to feat this data into the lua language server instead? I.e. especially the information of API objects and function and possible deprecation.

**Comment by Thaodan** [comment]
> > Maybe it is more IDE agnostic to bundle the type annotations as [LLS Addon](https://github.com/LuaLS/LLS-Addons) and distribute it as this? 🤔
> 
> I'm certainly entertaining the idea of bundling it as a LLS Addon for IDE agnostic support. I do need to find out how the vscode 3rd party extension (which has some extra features) would not conflict with the LLS addon

How different or what are these vscode specific features?

**Comment by Ketho** [comment]
Those additonal vscode features are not really essential, but still nice to have:
- code completion and hover tooltip for globalstrings
- hover tooltip for LE enums and cvars
- hover tooltip for events for the payload
- flavor information for API functions (mainline, cataclysm and vanilla)

**Comment by Thaodan** [comment]
Can those be implemented as a Lua-language-server extension? Lsp servers
can provide competition and docstrings so those parts are easy to
provide.


> hover tooltip for globalstrings to avoid polluting the global
> namespace

What do you mean by that?

Most of the things you mention should be doable as Lua-language-server
annotation:
https://luals.github.io/wiki/annotations/

**Comment by Ketho** [comment]
There are thousands of globalstrings (and a bunch of Lua Enums) which would show up when typing anything, when users barely even want to get completion for globalstrings so i'm using my own completion provider for those.

Yeah some are doable with annotations and docstrings like the flavor thing. But it does not support annotating events and cvars as they are just strings.

**Comment by Joshuahelmle** [comment]
Hey, has there been any updates to a LLS addons version for this?

I couldn't find anything, but since the Annotations are all avaiable in the repo it was straight forward to integrate it with LuaLS on nvim by just cloning the repo and pointing the luarc.json to the Annotations, maybe thats a worthy addittion to the readme or wiki? 
I assume the process will be the same for every ide/editor that supports LuaLS and therefore make it avaiable outside vscode too.
